src/detours.cpp

```diff
@@ -1838,76 +1919,16 @@ LONG WINAPI DetourTransactionCommitEx(_Out_opt_ PVOID **pppFailedPointer)
-        CONTEXT cxt;
-        cxt.ContextFlags = CONTEXT_CONTROL;
-
-#undef DETOURS_EIP
-
-#ifdef DETOURS_X86
-#define DETOURS_EIP         Eip
-#endif // DETOURS_X86
-
-#ifdef DETOURS_X64
-#define DETOURS_EIP         Rip
-#endif // DETOURS_X64
-
-#ifdef DETOURS_IA64
-#define DETOURS_EIP         StIIP
-#endif // DETOURS_IA64
-
-#ifdef DETOURS_ARM
-#define DETOURS_EIP         Pc
-#endif // DETOURS_ARM
-
-#ifdef DETOURS_ARM64
-#define DETOURS_EIP         Pc
-#endif // DETOURS_ARM64
-
-typedef ULONG_PTR DETOURS_EIP_TYPE;
-
-        if (GetThreadContext(t->hThread, &cxt)) {
-            for (o = s_pPendingOperations; o != NULL; o = o->pNext) {
-                if (o->fIsRemove) {
-                    if (cxt.DETOURS_EIP >= (DETOURS_EIP_TYPE)(ULONG_PTR)o->pTrampoline &&
-                        cxt.DETOURS_EIP < (DETOURS_EIP_TYPE)((ULONG_PTR)o->pTrampoline
-                                                             + sizeof(o->pTrampoline))
-                       ) {
-
-                        cxt.DETOURS_EIP = (DETOURS_EIP_TYPE)
-                            ((ULONG_PTR)o->pbTarget
-                             + detour_align_from_trampoline(o->pTrampoline,
-                                                            (BYTE)(cxt.DETOURS_EIP
-                                                                   - (DETOURS_EIP_TYPE)(ULONG_PTR)
-                                                                   o->pTrampoline)));
-
-                        SetThreadContext(t->hThread, &cxt);
-                    }
-                }
-                else {
-                    if (cxt.DETOURS_EIP >= (DETOURS_EIP_TYPE)(ULONG_PTR)o->pbTarget &&
-                        cxt.DETOURS_EIP < (DETOURS_EIP_TYPE)((ULONG_PTR)o->pbTarget
-                                                             + o->pTrampoline->cbRestore)
-                       ) {
-
-                        cxt.DETOURS_EIP = (DETOURS_EIP_TYPE)
-                            ((ULONG_PTR)o->pTrampoline
-                             + detour_align_from_target(o->pTrampoline,
-                                                        (BYTE)(cxt.DETOURS_EIP
-                                                               - (DETOURS_EIP_TYPE)(ULONG_PTR)
-                                                               o->pbTarget)));
-
-                        SetThreadContext(t->hThread, &cxt);
-                    }
-                }
-            }
-        }
-#undef DETOURS_EIP
+        detour_move_thread_instruction_pointer_if_needed(t->hThread);
     }
 
     // Restore all of the page permissions and flush the icache.
     HANDLE hProcess = GetCurrentProcess();
-    for (o = s_pPendingOperations; o != NULL;) {
+    for (o = s_pPendingOperations; o != NULL; o = o->pNext) {
         // We don't care if this fails, because the code is still accessible.
         DWORD dwOld;
         VirtualProtect(o->pbTarget, o->pTrampoline->cbRestore, o->dwPerm, &dwOld);
```

Comment:
Note that AppVerifier can also hook `VirtualProtect`, but I carefully reviewed the code for its hook and it never acquires any locks. I think back in 2014 when this change was originally made, the AppVerifier hook for `VirtualProtect` did acquire locks, so we moved the `VirtualProtect` calls to be after the `ResumeThread` calls. However, I didn't include that here since I don't think it is needed with the current version of AppVerifier's `VirtualProtect` hook. Also, for safety reasons, it is good to remove writability from executable pages before we resume threads.

The code for `VirtualProtect` itself can acquire a lock if we are restoring to read_execute and something has called `MmSecureVirtualMemory` with `PAGE_READWRITE`. However, this is very unlikely to happen in the case of Detours. As I understand it, `MmSecureVirtualMemory` with `PAGE_READWRITE` can be used by a driver if you pass it a user-mode buffer that it needs to fill in from kernel-mode. The API would ensure user-mode code can't switch the page to be read-only before the kernel-mode code gets a chance to write to the buffer. Executable code should not be used as the destination for a write from a driver, so I'm OK with a deadlock if this happens.